scilpy/reconst/multi_processes.py

```diff
@@ -0,0 +1,420 @@
+        if not np.isnan(shm_coeff[idx]).any():
+            odf = np.dot(shm_coeff[idx], B)
+            odf = odf.clip(min=0)
+            sum_odf = np.sum(odf)
+            max_odf = np.maximum(max_odf, sum_odf)
+            if sum_odf > 0:
+                rgb_map[idx] = np.dot(np.abs(sphere.vertices).T, odf)
+                rgb_map[idx] /= np.linalg.norm(rgb_map[idx])
+                rgb_map[idx] *= sum_odf
+            gfa_map[idx] = gfa(odf)
+            if gfa_map[idx] < gfa_thr:
+                global_max = max(global_max, odf.max())
+            elif np.sum(peak_indices[idx] > -1):
+                nufo_map[idx] = np.sum(peak_indices[idx] > -1)
+                afd_map[idx] = peak_values[idx].max()
+                afd_sum[idx] = np.sqrt(np.dot(shm_coeff[idx], shm_coeff[idx]))
+                qa_map = peak_values[idx] - odf.min()
+                global_max = max(global_max, peak_values[idx][0])
+
+    rgb_map /= max_odf
+    rgb_map *= 255
+    qa_map /= global_max
+
+    return chunk_id, nufo_map, afd_map, afd_sum, rgb_map, gfa_map, qa_map
+
+
+def maps_from_sh(shm_coeff, peak_dirs, peak_values, peak_indices, sphere,
+                 mask=None, gfa_thr=0, sh_basis_type='descoteaux07',
+                 nbr_processes=None):
+    """Computes maps from given SH coefficients and peaks
+
+    Parameters
+    ----------
+    shm_coeff : np.ndarray
+        Spherical harmonic coefficients
+    peak_dirs : np.ndarray
+        Peak directions
+    peak_values : np.ndarray
+        Peak values
+    peak_indices : np.ndarray
+        Peak indices
+    sphere : Sphere
+        The Sphere providing discrete directions for evaluation.
+    mask : np.ndarray, optional
+        If `mask` is provided, only the data inside the mask will be
+        used for computations.
+    gfa_thr : float, optional
+        Voxels with gfa less than `gfa_thr` are skipped for all metrics, except
+        `rgb_map`.
+        Default: 0
+    sh_basis_type : str, optional
+        Type of spherical harmonic basis used for `shm_coeff`. Either
+        `descoteaux07` or `tournier07`.
+        Default: `descoteaux07`
+    nbr_processes: int, optional
+        The number of subprocesses to use.
+        Default: multiprocessing.cpu_count()
+
+    Returns
+    -------
+    tuple of np.ndarray
+        nufo_map, afd_map, afd_sum, rgb_map, gfa, qa
+    """
+    sh_order = order_from_ncoef(shm_coeff.shape[-1])
+    B, _ = sh_to_sf_matrix(sphere, sh_order, sh_basis_type)
+
+    data_shape = shm_coeff.shape
+    if mask is not None:
+        shm_coeff = applymask(shm_coeff, mask)
+        peak_dirs = applymask(peak_dirs, mask)
+        peak_values = applymask(peak_values, mask)
+        peak_indices = applymask(peak_indices, mask)
+
+    nbr_processes = multiprocessing.cpu_count() if nbr_processes is None \
+        or nbr_processes < 0 else nbr_processes
+
+    npeaks = peak_values.shape[3]
+    shm_coeff = shm_coeff.ravel().reshape(np.prod(data_shape[0:3]), data_shape[3])
```

Comment:
Since you did that in 3 functions, maybe you could create a raveling/chunking function that takes as a input data_shape (4D) and a list of 3D or 4D array and return a list of list of array (in the same order).